doc/configuration.md

```diff
@@ -0,0 +1,170 @@
+# Custom Configuration Design for azure-init
+
+## Objective
+
+The azure-init custom configuration architecture enables dynamic and flexible management of various settings for virtual machines provisioned with the light-weight agent. Customizable settings include SSH, IMDS, provisioning media, azure proxy agent, wireserver, and telemetry.
+
+## Design
+
+The system supports default configurations with user-specified overrides via configuration files or CLI parameters.
+
+### Key Features
+
+- **Config Validation**: Validates user-provided settings against supported options, rejecting invalid values during deserialization.
+- **Built-in Defaults**: The system defines defaults directly in the code, eliminating the need for a separate default config file.
+- **Merging of Configurations**: Combines configuration sources hierarchically, applying defaults first, followed by file-based overrides, and finally CLI parameters for the highest precedence.
+
+## Config File Structure
+
+**Format**: TOML
+
+- Users can override default settings by supplying a single configuration file or multiple `.toml` files in a directory.
+
+### CLI Parameters
+
+Example: `--config /etc/azure-init/`
+
+## Configuration Hierarchy for `azure-init`
+
+Configuration can be set via a single file, a directory containing multiple files, or default values defined in the code.
+
+### Configuration Loading Order
+
+#### 1. CLI Override (`--config`)
+
+- The `--config` flag specifies the path to load the configuration from, which can be either a single file or a directory.
+  - **File:** If the path points to a file, only that file is loaded.
```

Comment:
From reading the code, this is not true. This file is still merged with the base azure-init.toml and also with defaults.
The order of merging part should be applicable to both files and directory merging.